proposals/p5168.md

```diff
@@ -0,0 +1,982 @@
+| `fn F[T:! I](x: T) {` ... `}`                                                       | `I` complete       | `F` complete                                                                           |                                                              |
+| `interface I;` <br> `class C;` <br> `class D(T:! I);` <br> `fn F(x: D(C));`         | `C impls I`        |                                                                                        | `I`, `C`?, `D`                                               |
+| `interface I;` <br> `class D(T:! I);` <br> `fn F[U:! type](x: D(U));`               |                    | `U impls I` <br> (implied constraint)                                                  | `I`, `D`, `F`                                                |
+
+> Whether classes need to be complete by the end of the file is not the subject
+> of this proposal.
+
+## Details
+
+### An `impl` implements a single interface
+
+We resolve
+[leads question #4566](https://github.com/carbon-language/carbon-lang/issues/4566)
+with two rules:
+
+-   Implementing an interface does not implement any of the interfaces it
+    extends.
+-   An impl declaration should implement a single interface.
+
+We allow facet type expressions to the right of `impl`...`as`, as long as that
+facet type corresponds to a single interface (ignoring interfaces it extends,
+which incidentally seems to
+[match Rust supertraits](https://stackoverflow.com/questions/75847426/why-do-we-need-a-separate-impl-for-a-supertrait)).
+This supports the use case of `Core.Add` actually being a named constraint
+defined in terms of `Core.AddWith`. This requires that the facet type is
+identified, at a minimum, when used in an `impl` declaration, so it can be
+resolved into the single interface actually being implemented.
+
+The "subsumption" use case of "when an interface X is implemented, Y will be
+implemented without the user having to write a separate impl definition" will
+instead be handled using
+[blanket implementations](/docs/design/generics/details.md#blanket-impl-declarations).
+There are a couple of variations:
+
+-   One interface is a strict superset of the other. In this case it would be a
+    lot less confusing/surprising if the interfaces use the same implementations
+    of functions that overlap. This is accomplished using a
+    [`final`](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#final-impl-declarations)
+    blanket implementation. For example, we will define
+    `final impl forall [T:! type, U:! ImplicitAs(T)] U as As(T)` and have
+    `As(T).Convert` forward to `ImplicitAs(T).Convert`. This way types will
+    either implement `ImplicitAs(T)` or `As(T)`, and will get an error if they
+    try to implement both.
```

Comment:
I feel like we discussed this, and maybe I just forgot the answer, so apologies for that.

But could we make this the meaning of `extend` in an interface, and have the other case be handled with `requires Self impls ...` and an impl?

Specifically, could we have `extend` synthesize the final blanket impl as suggested here, and be restricted to only be allowed in the case that we can do so?

Reply:
That is not something we discussed, but would be viable. It would replace future work to have a way to do this more easily, at the cost of restricting `extend` to interfaces in the same file. It would mean that the combination of `require Self impls`... and `extend api`... that is currently represented by `extend` would be harder to say, but not too bad once we accept the `extend api` proposal.